pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -405,80 +409,102 @@ func (f *fqdnController) deleteRuleSelectedPods(ruleID string) error {
+		ipWithTTLMap[ip] = ipMeta
+		minTimeToReQuery = earlierOf(ipMeta.expirationTime, minTimeToReQuery)
+	}
 
 	f.fqdnSelectorMutex.Lock()
 	defer f.fqdnSelectorMutex.Unlock()
-	oldDNSMeta, exist := f.dnsEntryCache[fqdn]
+
+	cachedDNSMeta, exist := f.dnsEntryCache[fqdn]
 	if exist {
-		mustCacheResponse = true
-		for ipStr := range responseIPs {
-			if _, ok := oldDNSMeta.responseIPs[ipStr]; !ok {
-				addressUpdate = true
-				break
+		// Data related to this domain exists in the cache.
+		// Besides presence of existing IPs, two scenarios : 1) May get New IPs 2) Some old IPs may be absent.
+
+		// check for new IPs and these new IPs need to be added with its new TTL value as received in response.
+		for ipStr, ipMeta := range newDNSresponseIPs {
+			if _, exist := cachedDNSMeta.responseIPs[ipStr]; !exist {
+				addIPToCache(ipStr, ipMeta)
 			}
 		}
-		for oldIPStr, oldIP := range oldDNSMeta.responseIPs {
-			if _, ok := responseIPs[oldIPStr]; !ok {
-				if oldDNSMeta.expirationTime.Before(time.Now()) {
-					// This IP entry has already expired and not seen in the latest DNS response.
-					// It should be removed from the cache.
+
+		for cachedIpStr, cachedIpMeta := range cachedDNSMeta.responseIPs {
+			if _, exist := newDNSresponseIPs[cachedIpStr]; !exist {
+				if cachedIpMeta.expirationTime.Before(currentTime) {
+					// this ip is expired and stale, remove it by not including it.
 					addressUpdate = true
 				} else {
-					// Add the unexpired IP entry to responseIP and update the lowest applicable TTL if needed.
-					responseIPs[oldIPStr] = oldIP
-					if oldDNSMeta.expirationTime.Before(recordTTL) {
-						recordTTL = oldDNSMeta.expirationTime
-					}
+					// It hasn't expired yet, so just retain it with its existing expirationTime.
+					addIPToCache(cachedIpStr, cachedIpMeta)
 				}
+			} else {
+				// This old IP also exists in current response, so update it with max time between received time and its old cached time.
+				expTime := laterOf(newDNSresponseIPs[cachedIpStr].expirationTime, cachedIpMeta.expirationTime)
+				addIPToCache(cachedIpStr, ipWithTTL{
+					ip:             cachedIpMeta.ip,
+					expirationTime: expTime,
+				})
 			}
 		}
+
 	} else {
+		// First time seeing this domain.
+		// check if this needs to be tracked, by checking its presence in the rules.
+		// If a FQDN policy had been applied then there must be rule records but because it's
+		// not in cache hence its FQDN:SelectorItem mapping may not be present.
 		for selectorItem := range f.selectorItemToRuleIDs {
 			// Only track the FQDN if there is at least one fqdnSelectorItem matching it.
 			if selectorItem.matches(fqdn) {
-				mustCacheResponse, addressUpdate = true, true
 				f.setFQDNMatchSelector(fqdn, selectorItem)
+				for ipStr, ipMeta := range newDNSresponseIPs {
+					addIPToCache(ipStr, ipMeta)
+				}
 			}
 		}
 	}
-	if mustCacheResponse {
+
+	if len(ipWithTTLMap) > 0 {
+		addressUpdate = true
```

Comment:
I don't think this is correct. Isn't `addressUpdate` supposed to be determined by whether FQDN <-> IP mappings have changed?

Reply:
@Dyanngg updated .